### PR TITLE
Add romfile option for NIC device

### PR DESCRIPTION
**Pull request by zhencliu**
Introduced a new VT param 'nic_romfile',
  To disable rom loading for all nics:
  nic_romfile = 'EMPTY_STRING'
  To disable rom loading for nic1
  nic_romfile_nic1 = 'EMPTY_STRING'

Signed-off-by: Zhenchao Liu <zhencliu@redhat.com>

ID: 2119345

### CODE REVIEW

**Comment by zhencliu** [comment]
Ran nic_hotplug.multi_pci.nic_virtio with:
nic_romfile = EMPTY_STRING
nic_romfile_nic1 = EMPTY_STRING
nic_romfile_hotplug_nic1 = EMPTY_STRING
All passed.

Some snippets of cmdline:
-device virtio-net-pci,mac=9a:2d:0f:d3:9f:b4,romfile=,id=idC26jOd,netdev=idhkhJqM,bus=pcie-root-port-3,addr=0x0

Send command: {'execute': 'device_add', 'arguments': OrderedDict([('id', 'idKegII5'), ('driver', 'virtio-net-pci'), ('netdev', 'idkIKiok'), ('mac', '9a:2d:0f:d3:9f:b5'), ('romfile', ''), ('bus', 'pcie_extra_root_port_0'), ('addr', '0x0')]), 'id': 'R62s1VQU'}

**Comment by zhencliu** [comment]
hi @PaulYuuu , as we discussed in #3491 , romfile is not a new param, and we also need to support hotplug.
hi @luckyh , I changed romfile to nic_romfile, I found that nic_romfile in __slots__ in utils_net.py is recognized and handled as a cartesian param currently. 
Would you please review the PR? Thanks.

**Comment by PaulYuuu** [comment]
As I said, `nic_extra_params` has the ability to handle this, for uncommon options, it's not necessary to introduce a new parameter for individual option.

```
nic_extra_params += "romfile=EMPTY_STRING"
ERROR: QMP command 'device_add' failed    (arguments: OrderedDict([('id', 'idM7hZ8i'), ('driver', 'virtio-net-pci'), ('netdev', 'idEpu3x2'), ('mac', '9a:51:67:4d:e2:0d'), ('romfile', '""'), ('bus', 'pcie_extra_root_port_0'), ('addr', '0x0')])
```

So after the empty string problem is fixed, this issue should be gone.

**Comment by zhencliu** [comment]
> As I said, `nic_extra_params` has the ability to handle this, for uncommon options, it's not necessary to introduce a new parameter for individual option.
> 
> ```
> nic_extra_params += "romfile=EMPTY_STRING"
> ERROR: QMP command 'device_add' failed    (arguments: OrderedDict([('id', 'idM7hZ8i'), ('driver', 'virtio-net-pci'), ('netdev', 'idEpu3x2'), ('mac', '9a:51:67:4d:e2:0d'), ('romfile', '""'), ('bus', 'pcie_extra_root_port_0'), ('addr', '0x0')])
> ```
> 
> So after the empty string problem is fixed, this issue should be gone.

hi @PaulYuuu , I should have a try with nic_extra_params when you made the comments, thanks for pointing this out.

hi @luckyh , It's easy(no code change) and I am OK with the nic_extra_params solution Yihuang mentioned, but I am a little concerned about the romfile code in VT at present, do you think we still need the code change?

**Comment by luckyh** [comment]
@zhencliu to fix the failure of CI Static checks, the word `romfile` has to be added into this [file](https://github.com/avocado-framework/avocado-vt/blob/master/spell.ignore).

**Comment by zhencliu** [comment]
> @zhencliu to fix the failure of CI Static checks, the word `romfile` has to be added into [this file](https://github.com/avocado-framework/avocado-vt/blob/master/spell.ignore).

hi @luckyh , cool, I used to use 'pylint: disable=xxx' to disable the spelling errors, thanks.

**Comment by zhencliu** [comment]
hi @luckyh , the change is done, now we can use romfile as an attribute in utils_net while use nic_romfile as a cartesian param, would you please make a double check?
hi @PaulYuuu , would you please make a double check?

I tested the following:
1. --customsparams='nic_romfile=EMPTY_STRING'
2. --customsparams='romfile=EMPTY_STRING'
both qemu and hotplug qmp cmdline can work fine, for 2, romfile will not affect the value setting of romfile attribute.

Thanks.

File: virttest/virt_vm.py [hunk 1]
```diff
@@ -937,6 +937,13 @@ def add_nic(self, **params):
         """
         if 'nic_name' not in params:
             params['nic_name'] = utils_misc.generate_random_id()
+
+        # 'romfile' is not a nic device cartesian param, remove it because it
+        # can overwrite the romfile attribute of a nic device, use 'nic_param'
+        # to set romfile attribute
+        if 'romfile' in params:
+            del params['romfile']
+
```
Comment (zhencliu) [inline virttest/virt_vm.py:946]: we don't need this code change, maybe.
if we don't set romfile in cfg, it should work fine.
what do you think @luckyh

Comment (luckyh) [reply virttest/virt_vm.py:946]: @zhencliu yes, agree. Let's remove it.

Comment (zhencliu) [reply virttest/virt_vm.py]: removed, thanks

**Review by luckyh — APPROVED** [review]
LGTM

**Review by PaulYuuu — APPROVED** [review]

**Comment by zhencliu** [comment]
Ran nic_hotplug.multi_pci.nic_virtio with the following settings:
1. nic_romfile_hotplug_nic2 = EMPTY_STRING
romfile is set to '' only when doing hotplug for hotplug_nic2 device
{'execute': 'device_add', 'arguments': OrderedDict([('id', 'idn6jhCw'), ('driver', 'virtio-net-pci'), ('netdev', 'idv2YqfR'), ('mac', '9a:29:ed:37:76:1b'), ('romfile', ''), ('bus', 'pcie_extra_root_port_1'), ('addr', '0x0')]), 'id': 'dHLQOCZV'}

2. nic_romfile = EMPTY_STRING
All nic devices' romfile is set to ''
-device virtio-net-pci,mac=9a:9d:07:cf:24:11,romfile=,id=idHjmbq2,netdev=idMiI1W1,bus=pcie-root-port-3,addr=0x0 
{'execute': 'device_add', 'arguments': OrderedDict([('id', 'idEB8CvR'), ('driver', 'virtio-net-pci'), ('netdev', 'idSwuCVI'), ('mac', '9a:9d:07:cf:24:12'), ('romfile', ''), ('bus', 'pcie_extra_root_port_0'), ('addr', '0x0')]), 'id': 'XkZ4dOlT'}
{'execute': 'device_add', 'arguments': OrderedDict([('id', 'idx5jzok'), ('driver', 'virtio-net-pci'), ('netdev', 'idLpbh5g'), ('mac', '9a:9d:07:cf:24:13'), ('romfile', ''), ('bus', 'pcie_extra_root_port_1'), ('addr', '0x0')]), 'id': 'g5Lmr9gc'}

3. without setting nic_romfile, romfile will not appear in cmdline.

All passed.

hi @luckyh , would you please help to merge it, thanks

**Comment by luckyh** [comment]
Thanks to all.

**Pull request merged**